src/platform/backends/qemu/qemu_mount_handler.cpp

```diff
@@ -38,54 +29,61 @@ namespace
 constexpr auto category = "qemu-mount-handler";
 } // namespace
 
-mp::QemuMountHandler::QemuMountHandler(const SSHKeyProvider& ssh_key_provider) : MountHandler(ssh_key_provider)
+namespace multipass
 {
-}
-
-void mp::QemuMountHandler::init_mount(VirtualMachine* vm, const std::string& target_path, const VMMount& vm_mount)
+QemuMountHandler::QemuMountHandler(VirtualMachine* vm, const SSHKeyProvider* ssh_key_provider, std::string target,
+                                   const VMMount& mount)
+    : MountHandler{vm, ssh_key_provider, target, mount}
 {
-    using St = VirtualMachine::State;
-    const auto skip_states = {St::off, St::stopped, St::suspended};
-
-    auto state = vm->current_state();
-    if (std::none_of(cbegin(skip_states), cend(skip_states), [&state](const auto& st) { return state == st; }))
-        throw std::runtime_error("Please shutdown virtual machine before defining native mount.");
+    if (!(qemu_vm = dynamic_cast<QemuVirtualMachine*>(vm)))
+        throw std::runtime_error{"QEMU mount must be used on a QEMU VM"};
```

Comment:
Why not receive a `QemuVirtualMachine` to enforce that at compilation time? The VMs make these, so you shouldn't need the RTTI. If some other caller wants to construct directly, they'd get the onus of ensuring the right type.

Reply:
Ah, yes, very good point. Thanks!

Reply:
Hi, @ricab! Unfortunately I have to address this point again, as I've run into some difficulties while writing tests. When testing the `QemuMountHandler`, I need to pass a mocked VM to the handler, for stuff like `vm->current_state()`. But if the constructor parameter is of type `QemuVirtualMachine*` instead of the base `VirtualMachine*`, I cannot pass a mock in a clean way. I tried to get around it with `reinterpret_cast`, but that brings other problems when the handler calls `QemuVirtualMachine` specific methods, like `modifiable_mount_args`, which I cannot mock. Of course, the previous `dynamic_cast<QemuVirtualMachine*>` was not a complete solution, since it won't be able to cast from `MockVirtualMachine*`, so it only moves the type problem to runtime. But then at least we could do the dynamic cast a little later, after some other checks that could be mocked. Any ideas or suggestions on how to approach this?

Reply:
I see two options: create a `MockQemuVirtualMachine` that inherits from `QemuVirtualMachine`, so that I can cleanly mock anything I need. Or pull the `QemuVirtualMachine` specific methods (just `modifiable_mount_args` as of now) into the base `VirtualMachine` and use the existing `MockVirtualMachine`. In my opinion, the second approach is kind of ugly. Every type of virtual machine will then inherit an unreasonable method for their specific implementation, and will end up throwing an exception when that method is called. I think the first approach would be cleaner, except that I would need to re-mock all the `VirtualMachine` methods. Do you know any nice multiple inheritance trick so that I can do something like
```cpp
struct MockQemuVirtualMachine : MockVirtualMachine, QemuVirtualMachine {
    // mock only QemuVirtualMachine specific stuff
}
```
so that it can be passed as a `QemuVirtualMachine*` but still use the mocked stuff from `MockVirtualMachine`?

Reply:
OK, discussed elsewhere, but a couple of lines of code to make the idea more concrete:

```cpp
template <typename VMClass = VM>
class MockVM : public VMClass
{
// Mock VM methods
};

class MockQemuVM : public MockVM<QemuVM>
{
// Mock QemuVM methods
};
```

Reply:
It works pretty nicely! The only minor inconvenience is that there are still several places that need to be modified with a `MockVirtualMachine<>` even though there is a default template parameter, like in the return type of functions or when used as a type in another template, like, `NiceMock<MockVirtualMachine<>>`.

So in order to avoid all that, I'm thinking to change the class name to something different and alias the concrete type with the default value to `MockVirtualMachine`
```cpp
template<typename T = VirtualMachine, typename = std::enable_if_t<std::is_base_of_v<VirtualMachine, T>>>
struct MockVirtualMachineT : public T {
...
}

using MockVirtualMachine = MockVirtualMachineT<>;
```

Reply:
Ah, and the other issue is that `QemuVirtualMachine` is declared as `final`, but I assume it's not a really big deal if it is removed.